cvmfs/cache_ram.cc

```diff
@@ -141,59 +206,156 @@ void RamCacheManager::CtrlTxn(
-  int64_t handle = CommitToKvStore(transaction);
-  if (handle < 0)
-    return handle;
-  ReadOnlyFd file_descriptor(handle, transaction->pos);
-
   WriteLockGuard guard(rwlock_);
-  return AddFd(file_descriptor);
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+  int64_t retval = CommitToKvStore(transaction);
+  if (retval < 0) {
+    LogCvmfs(kLogCache, kLogDebug,
+             "error while commiting transaction on %s: %s",
+             transaction->id.ToString().c_str(), strerror(-retval));
+    return retval;
+  }
+  LogCvmfs(kLogCache, kLogDebug, "open pending transaction for %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_openfromtxn);
+  return DoOpen(transaction->id);
 }
 
 
 int RamCacheManager::AbortTxn(void *txn) {
   Transaction *transaction = reinterpret_cast<Transaction *>(txn);
   if (transaction->buffer)
     free(transaction->buffer);
+  LogCvmfs(kLogCache, kLogDebug, "abort transaction %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_aborttxn);
   return 0;
 }
 
 
 int RamCacheManager::CommitTxn(void *txn) {
+  WriteLockGuard guard(rwlock_);
   Transaction *transaction = reinterpret_cast<Transaction *>(txn);
-  int retval = CommitToKvStore(transaction);
-  return retval;
+  perf::Inc(counters_.n_committxn);
+  return CommitToKvStore(transaction);
 }
 
 
 int64_t RamCacheManager::CommitToKvStore(Transaction *transaction) {
-  if (transaction->handle >= 0)
-    return transaction->handle;
+  kvstore::MemoryBuffer buf;
+  buf.address = transaction->buffer;
+  kvstore::MemoryKvStore *store;
+  if (transaction->expected_size == kSizeUnknown) {
+    buf.size = transaction->pos;
+    buf.address = realloc(buf.address, buf.size);
+    LogCvmfs(kLogCache, kLogDebug, "reallocating transaction on %s to %u B",
+             transaction->id.ToString().c_str(), buf.size);
+    if (!buf.address) {
+      LogCvmfs(kLogCache, kLogDebug, "realloc failed: %s", strerror(errno));
+      return -errno;
+    }
+  } else {
+    buf.size = transaction->size;
+  }
+  buf.object_type = transaction->object_type;
 
-  // TODO(jblomer): commit transaction in kv store
-  transaction->handle = 0;
-  if (transaction->buffer) {
-    free(transaction->buffer);
-    transaction->buffer = NULL;
+  if (buf.object_type == cache::CacheManager::kTypeVolatile) {
+    store = &volatile_entries_;
+  } else {
+    store = &regular_entries_;
+  }
+  if (buf.object_type == cache::CacheManager::kTypePinned ||
+      buf.object_type == cache::CacheManager::kTypePinned) {
+    buf.refcount = 1;
+  } else {
+    buf.refcount = 0;
+  }
+
+  int64_t regular_size = regular_entries_.GetUsed();
+  int64_t volatile_size = volatile_entries_.GetUsed();
+  int64_t overrun = regular_size + volatile_size + buf.size - max_size_;
+
+  if (overrun > 0) {
+    perf::Inc(counters_.n_overrun);
+    volatile_entries_.ShrinkTo(max((int64_t) 0, volatile_size - overrun));
```

Comment:
I think shrinking should be done down to some percent of the overall cache size (if possible).  Otherwise we'd constantly shrink once the cache is full.